### PR TITLE
Basic API

### PR DESCRIPTION
**Pull request by jackjwilliams**
Notes:

@seanohue I missed your comments about the keys validation, I will have to look at that.

I'm sure we can refactor this somehow to use the Data (and Area) class (I could have created a Bundle class as well, did not notice we weren't using one as of yet). I saw that Area.js hooks into updateTick, wasn't sure how to get around that as this is strictly for file / folder changes to "build out" areas / bundles (doesn't update GameState).

js-yaml does some weird stuff when saving yaml (it formats it different than the current items / rooms / npcs yaml files). I think it's still valid yaml, but ['item1', 'item2'] gets changed to 

- item 1
- item 2

I tried changing BundleManager to allow passing in a root src and bundles folder (to allow setting up "test" environments / bundles for writing tests), but the BundleManager and all subsequent core-* bundles are dependent on the src/ path. Wasn't sure how to proceed. Will setting up "test" environments / bundles require all core-* bundles be copied?

I'm holding off on the reboot endpoints and authentication for now as this PR is growing large, and that would be a significant refactor. We still only allow localhost, 127.0.0.1 and ::1.

#94

### CODE REVIEW

File: ranvier.json [hunk 1]
```diff
@@ -15,29 +19,35 @@
     "core-quests",
     "debug-commands"
   ],
-
+  "excludedBuilderBundles": [
```
Comment (shawncplus) [inline ranvier.json]: This isn't necessary. `bundles` already acts as a whitelist for data loaded into the game and any data loaded into the game should be loaded into the API

File: src/Data.js [hunk 2]
```diff
@@ -28,12 +28,41 @@ module.exports = {
     return parsers[ext](contents);
   },
 
+  saveFile: function(filepath, data, callback) {
+    if (!fs.existsSync(filepath)) {
+      throw new Error(`File [${filepath}] does not exist!`);
+    }
+
+    const deserializers = {
```
Comment (shawncplus) [inline src/Data.js]: Shouldn't these be called serializers?

Comment (shawncplus) [inline src/Data.js]: Also, you have trailing whitespace :P

**Comment by shawncplus** [comment]
This is a pretty big change so I'm just kind of trusting that the nuts & bolts of this works. I have to say I'm not a fan of the whole `FileManagement`/`*Folder` naming convention as it shouldn't matter to anyone except the classes loading the bundle data themselves that data is coming from the filesystem. If you want to change that now I would appreciate it but I'm not going to raise hell over it.

File: src/Data.js (hunk 2 above)
Comment (jackjwilliams) [inline src/Data.js]: image

**Pull request merged**